File: sdks/python/src/novelcraft/sdk/block_source.py
"""The block source APIs.

The block source API is the API that can be used to get information about
blocks.
"""
from __future__ import annotations

import math
from abc import ABC, abstractmethod
from typing import Dict

from .block import IBlock, Block
from .position import Position
from .section import Section


class IBlockSource(ABC):
    """The block source interface.

    The block source interface is the interface that can be used to get
    information about blocks.
    """
    @abstractmethod
    def __getitem__(self, position: Position[int]) -> IBlock:
        """Gets a block.

        Args:
            position (Position[int]): The position of the block.

        Returns:
            The block.
        """
        raise NotImplementedError()


class BlockSource(IBlockSource):
    _DEFAULT_BLOCK_ID = -1

    def __init__(self):
        super().__init__()

        self._section_dictionary: Dict[Position[int], Section] = {}

    def __getitem__(self, position: Position[int]) -> Block | None:
        section = self._get_section(position)
        if section is None:
            return None

        return section[Position[int](position.x - section.get_position().x,
                                position.y - section.get_position().y,
                                position.z - section.get_position().z)]

    def __setitem__(self, position: Position[int], block: Block) -> None:
        section = self._get_section(position)
        if section is None:
            section = Section(
                position, [BlockSource._DEFAULT_BLOCK_ID] * 4096)
            self.add_section(section)

        section[Position[int](position.x - section.get_position().x,
                         position.y - section.get_position().y,
                         position.z - section.get_position().z)] = block

    def add_section(self, section: Section) -> None:
        self._section_dictionary[section.get_position()] = section

    def clear(self) -> None:
        self._section_dictionary.clear()

    def remove_section(self, position: Position[int]) -> None:
        position = Position[int](16 * int(math.floor(position.x / 16)),
                            16 * int(math.floor(position.y / 16)),
                            16 * int(math.floor(position.z / 16)))

        self._section_dictionary.pop(position, None)

    def _get_section(self, position: Position[int]) -> Section | None:
        position = Position[int](16 * int(math.floor(position.x / 16)),
                            16 * int(math.floor(position.y / 16)),
                            16 * int(math.floor(position.z / 16)))

        return self._section_dictionary.get(position, None)

File: sdks/dotnet/src/BlockSource.cs
namespace NovelCraft.Sdk;

internal class BlockSource : IBlockSource {
  public IBlock? this[IPosition<int> position] {
    get {
      var section = GetSection(position);
      if (section is null) {
        return null;
      }

      return section[new Position<int>(position.X - section.Position.X, position.Y - section.Position.Y, position.Z - section.Position.Z)];
    }

    set {
      if (value is null) {
        throw new ArgumentNullException(nameof(value));
      }

      var section = GetSection(position);
      if (section is null) {
        throw new ArgumentException("The section is not loaded.");
      }

      section[new Position<int>(position.X - section.Position.X, position.Y - section.Position.Y, position.Z - section.Position.Z)] = value;
    }
  }

  public int Count => _sectionDictionary.Count * 16 * 16 * 16;

  private Dictionary<IPosition<int>, Section> _sectionDictionary = new();


  public void AddSection(Section section) {
    _sectionDictionary[section.Position] = section;
  }

  public void Clear() {
    _sectionDictionary.Clear();
  }

  public void RemoveSection(IPosition<int> position) {
    // Check if the position is valid.
    if (position.X % 16 != 0 || position.Y % 16 != 0 || position.Z % 16 != 0) {
      throw new ArgumentException("The position should be multiples of 16.");
    }

    if (_sectionDictionary.ContainsKey(position)) {
      _sectionDictionary.Remove(position);
    }
  }

  private Section? GetSection(IPosition<int> position) {
    Position<int> sectionPosition = new Position<int> {
      X = 16 * (int)Math.Floor(position.X / 16.0),
      Y = 16 * (int)Math.Floor(position.Y / 16.0),
      Z = 16 * (int)Math.Floor(position.Z / 16.0)
    };

    return _sectionDictionary.ContainsKey(sectionPosition) ? _sectionDictionary[sectionPosition] : null;
  }
}